plugins/terraform/layer0/resource_layer0_environment.go

```diff
@@ -1,10 +1,14 @@
-package main
+package layer0
 
 import (
 	"log"
 
 	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/quintilesims/layer0/api/provider/aws"
+	"github.com/quintilesims/layer0/client"
+	"github.com/quintilesims/layer0/common/config"
 	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
 )
 
 func resourceLayer0Environment() *schema.Resource {
```

Comment:
The [Exists](https://www.terraform.io/docs/plugins/provider.html#exists) Resource CRUD operation needs to be implemented here.

I created a simple environment with terraform (v0.11)
```
provider "layer0" {
  endpoint        = "${var.endpoint}"
  token           = "${var.token}"
  skip_ssl_verify = true
}

resource "layer0_environment" "demo" {
  name      = "demo"
  min_count = 1
  os        = "linux"
  size      = "t2.micro"
}
```
Then I removed this environment from the CLI. Running `terraform refresh` yields a 500 response from the layer0 API due to the security group of the environment not existing. It seems that this is due to `Exists` not being present within the Resouce's operations.

Reply:
We actually don't need that method, as it uses `Read` when `Exist` isn't implemenated (this is what was used in previous iterations). However, you have found a bug in the API: For the Terraform code to determine when an entity does not exist, we explicitly check that the Layer0 API returns a `<Entity>DoesNotExist` error code. This needs to happen when the environment's SG does not exist.